[quant] Add default symmetric qat qconfig for qnnpack

digantdesai opened the pull request:
Summary:
* This is the default symmetric qat qconfigs for qnnpack.
* Support for symmetric quantization is not available from other backends.
* Observers are similar to symmetric PTQ qconfigs for qnnpack.

Test Plan:
* Added new tests `test_conv_linear_symm` under `quantization.eager.test_quantize_eager_qat.TestQuantizeEagerQAT`, and `test_linear_bn_symm_numerics` under `quantization.eager.test_quantize_eager_qat.TestQuantizeEagerQATNumerics`
    * Only for qnnpack qengine
    * Skipping this test if XNNPACK is not available

Differential Revision: D34804808

jerryzh168 commented on test/quantization/eager/test_quantize_eager_qat.py:
```diff
@@ -340,6 +343,37 @@ def checkQuantized(model):
                 checkQuantized(model)
 
+    @skipIfNoXNNPACK
+    def test_conv_linear_symm(self):
```
do we need to add a test for numerics as well?

digantdesai replied:
Added! Thanks!

jerryzh168 commented on test/quantization/eager/test_quantize_eager_qat.py:
```diff
@@ -51,6 +52,8 @@
     override_qengines,
 )
 
+from torch.testing._internal.common_utils import skipIfNoXNNPACK
```
I thought we are going to hide xnnpack entirely, are we still exposing this in torch.backends?

digantdesai replied:
yes, that's the goal. I wish I didn't have to add this but for OSS CI there are some builds which don't have USE_XNNPACK, in those cases this test will fail. 

Until XNNPACK is optional we have to do this.

jerryzh168 reviewed (approved):
looks good, had some comments inline

jerryzh168 commented:
can we update the summary and test plan here as well?

JacobSzwejbka committed:
[Pytorch edge] Hackily fix qnnpack only model issue

Differential Revision: D33927201

fbshipit-source-id: cdaf95a4527a7e4b5070c32538217db6dbfbda22

digantdesai committed:
[quant] Add default symmetric qat qconfig for qnnpack (#74507)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/74507

* This is the default symmetric qat qconfigs for qnnpack.
* Support for symmetric quantization is not available from other backends.
* Observers are similar to symmetric PTQ qconfigs for qnnpack.

Reviewed By: jerryzh168

Differential Revision: D34804808

fbshipit-source-id: a880cc2f8043188b2c55b1a481292544d2e62828
